README.md
# Stork

TEST

<img align="right" src="/doc/static/stork-square-200px.png">

Stork is an open source ISC project providing a monitoring application and dashboard for 
ISC Kea DHCP and (eventually) ISC BIND 9. A limited configuration management for Kea
is available and is expected to grow substantially in the near future.

The project is currently in rapid development, with monthly releases rolling out new features.
See [wiki pages](https://gitlab.isc.org/isc-projects/stork/-/wikis/home) for useful
links to download page, release notes, self-guided demo, screenshots and much more.

For details, please see the [Stork Administrator Reference Manual](https://stork.readthedocs.io) 
or the [Stork wiki](https://gitlab.isc.org/isc-projects/stork/-/wikis/home).

# Build Instructions

The easiest way to run Stork is to install it using
[RPM and deb packages](https://stork.readthedocs.io/en/latest/install.html#installing-from-packages).
The second easiest way is to use Docker (`rake demo:up` or `./stork-demo.sh`). However, it is
possible to run Stork without Docker. See the Installation section of the Stork ARM.

# Reporting Issues

Please use the issue tracker on [ISC's GitLab](https://gitlab.isc.org/isc-projects/stork/-/issues)
to report issues and submit feature requests.

# Getting Involved

We have monthly development releases. If you'd like to get involved, feel free to subscribe to the
[stork-dev mailing list](https://lists.isc.org/mailman/listinfo/stork-dev) or look
at the [Stork project page](https://gitlab.isc.org/isc-projects/stork).
We're also on [GitHub](https://github.com/isc-projects/stork).

If you have a patch to send, by far the best way is to submit a
[merge request (MR) on GitLab](https://gitlab.isc.org/isc-projects/stork/-/merge_requests).
Stork developers use this system daily and you may expect a reasonably quick response.
The second alternative is to submit a [pull request (PR) on GitHub](https://github.com/isc-projects/stork/pulls).
This will also work, but this system is not monitored, so expect a delayed response.

# Screenshots

An example front page of the dashboard looks like this:
![Stork dashboard](https://gitlab.isc.org/isc-projects/stork/-/wikis/uploads/22cf367aedaaad3ac8e42d066595dd7b/dashboard-1.1.png)

Many more Stork screenshots are available on the [Screenshots gallery](https://gitlab.isc.org/isc-projects/stork/-/wikis/Screenshots).

# Prometheus and Grafana

Stork provides support for statistics export in Prometheus format, which can then easily be shown in Grafana.

An example of Kea dashboard in Grafana, displaying data exported with Stork: 
![grafana-kea4](https://gitlab.isc.org/isc-projects/stork/-/wikis/uploads/97468f53d07c1b6eda7035c30fbd4de3/grafana-kea4.png)

BIND9 dashboard in Grafana, displaying data exported with Stork:
![grafana-bind2](https://gitlab.isc.org/isc-projects/stork/-/wikis/uploads/6673c0a19962c535bf7e47d9fd0f46e5/grafana-bind2.png)




ChangeLog.md
* 324 [func] slawek

    Added a support for BIND 9 running in the chroot mode.
    (Gitlab #974)

* 323 [build] razvan

    Updated Kea version to 2.4.0 in demo and system tests.
    (Gitlab #995)

Stork 1.12.0 released on 2023-08-02.

* 322 [func] slawek

    Added a configuration option to handle requests directed to URL
    subdirectory (instead of URL root). Included an example Apache configuration
    file.
    (Gitlab #1039)

* 321 [build] slawek

    Refactored and formatted to the Python codebase to reach the highest
    quality rate from the linter.
    (Gitlab #1044)

* 320 [bug] slawek

    Fixed a bug in the script installing the Stork agent using a server token.
    This bug caused installation errors due to wrong paths to the deployed
    packages. The script no longer requires all Stork agent package formats to
    be deployed in the Stork server, and it will use the available formats for
    the matching operating systems where the Stork agent is installed.
    (Gitlab #932)

* 319 [func] marcin

    Created shared network view in the WebUI that contains shared network
    details, including DHCP parameters and options.
    (Gitlab #1119)

* 318 [func] slawek

    Added a new checker to verify if the Kea Control Agent configuration
    includes the control sockets. Added a descriptive message about possibly
    missing control socket configuration of the Kea DHCP daemons.
    (Gitlab #1045)

* 317 [bug] slawek

    Fixed bug with detecting RNDC key if the -c flag is not used to set the
    config path in BIND 9.
    (Gitlab #1057)

* 316 [func] slawek

    Refactored the log messages produced on the Stork agent initialization to be
    more straightforward and user-friendly.
    (Gitlab #1056)

* 315 [bug] slawek

    Enabled gzip compression for all communication between the Stork server and
    agent. It fixes handling a big response of named statistics endpoint.
    Stork accepts payloads (i.e., responses from Kea and BIND 9 endpoints) up
    to 40MiB uncompressed size.
    (Gitlab #1059)

* 314 [bug] slawek

    Fixed a UI problem that caused the IPv6 subnet bars to be unreadable for
    long addresses.
    (Gitlab #1060)

* 313 [bug] slawek

    Fixed a rare crash occurring when the state puller schedules the config
    review for an existing daemon. Set the focus on the login page to the first
    input. Shrank the width of the lease user context viewer to its content.
    Fixed a problem with the DHCP identifier button's width.
    (Gitlab #1053)

* 312 [build] slawek

    Refactored mock file suffixes. The mocks no longer need to be manually
    registered in the Rake file.
    (Gitlab #1006)

* 311 [func] marcin

    Subnet view contains DHCP parameters and options.
    (Gitlab #953)

* 310 [func] marcin

    Filter host reservations by subnet ID.
    (Gitlab #1058)

* 309 [func] marcin

    A tab with the subnet details is opened after clicking on the subnet.
    (Gitlab #931)

* 308 [doc] tomek

    The Stork Developer's Guide is now a stand-alone document. You can
    build it using rake `build:devguide` command.
    (Gitlab #786)

* 307 [bug] marcin

    Fixed the simulator in Stork demo.
    (Gitlab #1054)

Stork 1.11.0 released on 2023-06-07.

* 306 [bug] slawek

    Fixed propagating database password to the session manager.
    (Gitlab #1018)

* 305 [bug] slawek

    Fixed a problem with improperly using the rndc command that prevented
    detecting BIND 9 daemons if the RNDC key was specified outside the default
    RNDC key file. Fixed a rare bug when the Stork server crashes if an error
    occurs on a particular stage of BIND 9 detection.
    (Gitlab #1031)

* 304 [bug] slawek

    Fixed a bug in handling the empty database host that caused user login
    failures in the UI. Fixed rejecting the database usernames containing a
    dash. Fixed printing of the error messages related to the database
    authentication problems.
    (Gitlab #1022)

* 303 [func] razvan

    The lease search page can now display the content of the lease user
    context.
    (Gitlab #999)

* 302 [bug] marcin

    Fixed a bug in the Stork server that resulted in temporarily holding wrong
    DHCP options for an edited host reservation. This problem occurred when
    different options were specified for different servers owning the same
    host reservation. In addition, fixed a similar bug in specifying different
    client classes for different servers.
    (Gitlab #1017)

* 301 [bug] slawek

    Stork resolves now the include statements in the Kea configuration with any
    file extension.
    (Gitlab #1036)

* 300 [bug] slawek

    Fixed a bug in one of the config review checkers that caused a server crash
    if the Kea configuration contained a subnet prefix with a zero mask.
    (Gitlab #1024)

* 299 [bug] marcin

    Fixed a bug in the Stork server that could sometimes cause a modification
    of the timestamps of various records in the database during the update.
    This bug is unlikely to have a real impact on the users.
    (Gitlab #1007)

Stork 1.10.0 released on 2023-04-05.

* 298 [func] marcin

    New and updated host reservations are instantly visible in Stork after
    submitting the form.
    (Gitlab #996)

* 297 [build] slawek

    Fixed the security vulnerabilities reported by the Github Dependabot and
    updated dependencies including the Angular, PrimeNG, GoSwagger and
    OpenAPI Generator.
    (Gitlab #981)

* 296 [bug] slawek

    Fixed the path traversal vulnerability that allowed everyone to check the
    existence of any file on the filesystem.
    (Gitlab #987)

* 295 [bug] slawek

    Fixed fetching the authorization keys from BIND 9 configuration. The key
    value is visible only for super-administrators.
    (Gitlab #997)

* 294 [build] slawek

    Changed the executable paths configured in the default SystemD service
    files to absolute.
    (Gitlab #972)

* 293 [bug] slawek


backend/util/util.go
package storkutil

import (
	"bytes"
	"encoding/hex"
	"fmt"
	"os"
	"os/exec"
	"path"
	"path/filepath"
	"reflect"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"time"

	"github.com/pkg/errors"
	log "github.com/sirupsen/logrus"
	"golang.org/x/term"
)

func UTCNow() time.Time {
	return time.Now().UTC()
}

// Returns URL of the host with port.
func HostWithPortURL(address string, port int64, secure bool) string {
	protocol := "http"
	if secure {
		protocol = "https"
	}
	return fmt.Sprintf("%s://%s:%d/", protocol, address, port)
}

// Parses URL into host and port.
func ParseURL(url string) (host string, port int64, secure bool) {
	pattern := regexp.MustCompile(`https{0,1}:\/\/\[{1}(\S+)\]{1}(:([0-9]+)){0,1}`)
	m := pattern.FindStringSubmatch(url)

	if len(m) == 0 {
		pattern := regexp.MustCompile(`https{0,1}:\/\/([^\s\:\/]+)(:([0-9]+)){0,1}`)
		m = pattern.FindStringSubmatch(url)
	}

	if len(m) > 1 {
		host = m[1]
	}

	if len(m) > 3 {
		p, err := strconv.Atoi(m[3])
		if err == nil {
			port = int64(p)
		}
	}

	secure = strings.HasPrefix(url, "https://")

	// Set default ports
	if port == 0 {
		switch {
		case strings.HasPrefix(url, "http://"):
			port = 80
		case strings.HasPrefix(url, "https://"):
			port = 443
		}
	}

	return host, port, secure
}

// Formats provided string of hexadecimal digits to MAC address format
// using colon as separator. It returns formatted string and a boolean
// value indicating if the conversion was successful.
func FormatMACAddress(identifier string) (formatted string, ok bool) {
	// Check if the identifier is already in the desired format.
	identifier = strings.TrimSpace(identifier)
	pattern := regexp.MustCompile(`^[0-9A-Fa-f]{2}((:{1})[0-9A-Fa-f]{2})*$`)
	if pattern.MatchString(identifier) {
		// No conversion required. Return the input.
		return identifier, true
	}
	// We will have to convert it, but let's first check if this is a valid identifier.
	if !IsHexIdentifier(identifier) {
		return "", false
	}
	// Remove any colons and whitespace.
	replacer := strings.NewReplacer(" ", "", ":", "")
	numericOnly := replacer.Replace(identifier)
	for i, character := range numericOnly {
		formatted += string(character)
		// Divide the string into groups with two digits.
		if i > 0 && i%2 != 0 && i < len(numericOnly)-1 {
			formatted += ":"
		}
	}
	return formatted, true
}

// Detects if the provided string is an identifier consisting of
// hexadecimal digits and optionally whitespace or colons between
// the groups of digits. For example: 010203, 01:02:03, 01::02::03,
// 01 02 03 etc. It is useful in detecting if the string comprises
// a DHCP client identifier or MAC address.
func IsHexIdentifier(text string) bool {
	pattern := regexp.MustCompile(`^[0-9A-Fa-f]{2}((\s*|:{0,2})[0-9A-Fa-f]{2})*$`)
	return pattern.MatchString(strings.TrimSpace(text))
}

func SetupLoggingLevel() {
	// Setup logging level.
	//
	// If the STORK_LOG_LEVEL is specified and has valid name in it, use
	// that level. If not specified or has garbage, use the default (INFO).
	if value, ok := os.LookupEnv("STORK_LOG_LEVEL"); ok {
		levels := map[string]log.Level{
			"DEBUG": log.DebugLevel,
			"INFO":  log.InfoLevel,
			"WARN":  log.WarnLevel,
			"ERROR": log.ErrorLevel,
		}

		if level, ok := levels[value]; ok {
			// STORK_LOG_LEVEL specified, setting logging level to whatever was specified.
			log.SetLevel(level)
		} else {
			// Invalid level. Let's initialize logging first and then log a
			// warning about it.
			log.SetLevel(log.InfoLevel)
			log.Warnf("STORK_LOG_LEVEL has invalid log level: %s, ignoring.", value)
		}
	} else {
		// STORK_LOG_LEVEL not specified, using default logging level (INFO).
		log.SetLevel(log.InfoLevel)
	}
}

func SetupLogging() {
	// Normalizes the color environment variables from the standard Stork
	// convention.
	variables := []string{"CLICOLOR_FORCE", "CLICOLOR"}
	for _, variable := range variables {
		if value, ok := os.LookupEnv(variable); ok {
			if strings.ToLower(value) == "true" {
				os.Setenv(variable, "1")
			} else if strings.ToLower(value) == "false" {
				os.Setenv(variable, "0")
			}
		}
	}

	SetupLoggingLevel()

	log.SetOutput(os.Stdout)
	log.SetReportCaller(true)
	log.SetFormatter(&log.TextFormatter{
		EnvironmentOverrideColors: true,
		FullTimestamp:             true,
		ForceQuote:                true,
		TimestampFormat:           "2006-01-02 15:04:05",
		// TODO: do more research and enable if it brings value
		// PadLevelText: true,
		// FieldMap: log.FieldMap{
		// 	FieldKeyTime:  "@timestamp",
		// 	FieldKeyLevel: "@level",
		// 	FieldKeyMsg:   "@message",
		// },
		CallerPrettyfier: func(f *runtime.Frame) (string, string) {
			// Grab filename and line of current frame and add it to log entry
			_, filename := path.Split(f.File)
			return "", fmt.Sprintf("%20v:%-5d", filename, f.Line)
		},
	})
}

// Helper code for mocking os/exec stuff... pathetic.
type Commander interface {
	Output(string, ...string) ([]byte, error)
}

type RealCommander struct{}

func (c RealCommander) Output(command string, args ...string) ([]byte, error) {
	return exec.Command(command, args...).Output()
}

// Convert bytes to hex string.
func BytesToHex(bytesArray []byte) string {
	var buf bytes.Buffer
	for _, f := range bytesArray {
		fmt.Fprintf(&buf, "%02X", f)
	}
	return buf.String()
}

// Convert a string of hexadecimal digits to bytes array.
func HexToBytes(hexString string) []byte {
	hexString = strings.ReplaceAll(hexString, ":", "")
	decoded, _ := hex.DecodeString(hexString)
	return decoded
}

func GetSecretInTerminal(prompt string) string {
	// Prompt the user for a secret
	fmt.Print(prompt)
	pass, err := term.ReadPassword(0)
	fmt.Print("\n")

	if err != nil {
		log.Fatal(err.Error())
	}
	return string(pass)
}

// Read a file and resolve all include statements.
func ReadFileWithIncludes(path string) (string, error) {
	parentPaths := map[string]bool{}
	return readFileWithIncludes(path, parentPaths)
}

// Recursive function to read a file and resolve all include statements.
func readFileWithIncludes(path string, parentPaths map[string]bool) (string, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		log.Warnf("Cannot read file: %+v", err)
		err = errors.Wrap(err, "cannot read file")
		return "", err
	}

	text := string(raw)

	// Include pattern definition:
	// - Must start with prefix: <?include
	// - Must end with suffix: ?>
	// - Path may be relative to parent file or absolute
	// - Path must be escaped with double quotes
	// - May to contains spacing before and after the path quotes
	// - Path must contain ".json" extension
	// Produce two groups: first for the whole statement and second for path.
	includePattern := regexp.MustCompile(`<\?include\s*\"([^"]+\.json)\"\s*\?>`)
	matchesGroupIndices := includePattern.FindAllStringSubmatchIndex(text, -1)
	matchesGroups := includePattern.FindAllStringSubmatch(text, -1)

	// Probably never met
	if (matchesGroupIndices == nil) != (matchesGroups == nil) {
		return "", errors.New("include statement recognition failed")
	}

	// No matches - nothing to expand
	if matchesGroupIndices == nil {
		return text, nil
	}

	// Probably never met
	if len(matchesGroupIndices) != len(matchesGroups) {
		return "", errors.New("include statement recognition asymmetric")
	}

	// The root directory for includes
	baseDirectory := filepath.Dir(path)

	// Iteration from the end to keep correct index values because when the pattern
	// is replaced with an include content the positions of next patterns are shifting
	for i := len(matchesGroupIndices) - 1; i >= 0; i-- {
		matchedGroupIndex := matchesGroupIndices[i]
		matchedGroup := matchesGroups[i]

		statementStartIndex := matchedGroupIndex[0]
		matchedPath := matchedGroup[1]
		matchedStatementLength := len(matchedGroup[0])
		statementEndIndex := statementStartIndex + matchedStatementLength

		// Include path may be absolute or relative to a parent file
		nestedIncludePath := matchedPath
		if !filepath.IsAbs(nestedIncludePath) {
			nestedIncludePath = filepath.Join(baseDirectory, nestedIncludePath)
		}
		nestedIncludePath = filepath.Clean(nestedIncludePath)

		// Check for infinite loop
		_, isVisited := parentPaths[nestedIncludePath]
		if isVisited {
			err := errors.Errorf("detected infinite loop on include '%s' in file '%s'", matchedPath, path)
			return "", err
		}

		// Prepare the parent paths for the nested level
		nestedParentPaths := make(map[string]bool, len(parentPaths)+1)
		for k, v := range parentPaths {
			nestedParentPaths[k] = v
		}
		nestedParentPaths[nestedIncludePath] = true

		// Recursive call
		content, err := readFileWithIncludes(nestedIncludePath, nestedParentPaths)
		if err != nil {
			return "", errors.Wrapf(err, "problem with inner include: '%s' of '%s': '%s'", matchedPath, path, nestedIncludePath)
		}

		// Replace include statement with included content
		text = text[:statementStartIndex] + content + text[statementEndIndex:]
	}

	return text, nil
}

// Check if it is possible to create a file
// with the provided filename.
func IsValidFilename(filename string) bool {
	if strings.ContainsAny(filename, "*") {
		return false
	}
	file, err := os.CreateTemp("", filename+"*")
	if err != nil {
		return false
	}
	file.Close()
	os.Remove(file.Name())
	return true
}

// Returns a string comprising a count and a noun in the plural or
// singular form, depending on the count. The third parameter is a
// postfix making the plural form.
func FormatNoun(count int64, noun, postfix string) string {
	formatted := fmt.Sprintf("%d %s", count, noun)
	if count != 1 && count != -1 {
		formatted += postfix
	}
	return formatted
}

// Check if the interface is nil pointer - (*T)(nil). It is helpful
// in the functions that accept an interface type. If the real
// type is a pointer to a struct that implements the interface,
// then standard nil checking (with == operator) always returns
// false even if the pointer is nil. It isn't a big problem in
// most cases because you can call the interface methods on the
// nil interface (the receiver will be nil). It's dangerous if
// you use the type composition. If you don't override the methods
// in derived types, then all calls are piped to the base type.
// But if you try to do this on nil interface, then GO panics.
// You cannot prevent it by standard nil checking in the functions
// that use the interface as an argument type. You have to use this
// helper function. Very confusing.
//
// Warning! If you need to use this function, then probably
// your code is inconsistent. It would be best if you didn't
// allow that nil pointers will be cast to interface{}.
//
// Source: https://stackoverflow.com/a/50487104 .
// See: https://groups.google.com/g/golang-nuts/c/wnH302gBa4I
func IsNilPtr(obj interface{}) bool {
	return obj == nil || reflect.ValueOf(obj).Kind() == reflect.Ptr && reflect.ValueOf(obj).IsNil()
}

// Checks if the specified value is a whole number.
func IsWholeNumber(value interface{}) bool {
	if value == nil {
		return false
	}
	valueType := reflect.TypeOf(value)
	switch valueType.Kind() {
	case reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Int, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uint:
		return true
	default:
		return false
	}
}
